src/parquet/arrow/writer.cc

```diff
@@ -389,19 +391,18 @@ NULLABLE_BATCH_FAST_PATH(DoubleType, ::arrow::DoubleType, double)
 //   ArrowType::c_type to ParquetType::c_type
 template <>
 Status FileWriter::Impl::TypedWriteBatch<BooleanType, ::arrow::BooleanType>(
-    ColumnWriter* column_writer, const Array* array, int64_t offset, int64_t num_values,
-    int64_t num_levels, const int16_t* def_levels, const int16_t* rep_levels) {
-  DCHECK((offset + num_values) <= array->length());
-  RETURN_NOT_OK(data_buffer_.Resize(num_values));
-  auto data = static_cast<const BooleanArray*>(array);
+    ColumnWriter* column_writer, const std::shared_ptr<Array>& array, int64_t num_levels,
+    const int16_t* def_levels, const int16_t* rep_levels) {
+  RETURN_NOT_OK(data_buffer_.Resize(array->length()));
+  auto data = static_cast<const BooleanArray*>(array.get());
   auto data_ptr = reinterpret_cast<const uint8_t*>(data->data()->data());
   auto buffer_ptr = reinterpret_cast<bool*>(data_buffer_.mutable_data());
   auto writer = reinterpret_cast<TypedColumnWriter<BooleanType>*>(column_writer);
 
   int buffer_idx = 0;
-  for (int i = 0; i < num_values; i++) {
-    if (!data->IsNull(offset + i)) {
-      buffer_ptr[buffer_idx++] = BitUtil::GetBit(data_ptr, offset + i);
+  for (int i = 0; i < data->length(); i++) {
+    if (!data->IsNull(i)) {
+      buffer_ptr[buffer_idx++] = BitUtil::GetBit(data_ptr, array->offset() + i);
```

Comment:
Now that `array` is `shared_ptr<Array>`, this may affect performance (the compiler may not be able to reason about whether `array->offset()` is a loop invariant